Anomalies collector (#10060)
netdata/netdata: collectors/python.d.plugin/anomalies/README.md

_**Note**: If you are unsure about any of the below configuration options then it's best to just ignore all this and leave the `anomalies.conf` file alone to begin with. Then you can return to it later if you would like to tune things a bit more once the collector is running for a while and you have a feeling for its performance on your node._

Edit the `python.d/anomalies.conf` configuration file using `edit-config` from the your agent's [config
directory](https://learn.netdata.cloud/guides/step-by-step/step-04#find-your-netdataconf-file), which is usually at `/etc/netdata`.

```bash
cd /etc/netdata   # Replace this path with your Netdata config directory, if different
sudo ./edit-config python.d/anomalies.conf
```

The default configuration should look something like this. Here you can see each parameter (with sane defaults) and some information about each one and what it does.

```yaml
# what host to pull data from.

> host: '127.0.0.1:19999'

Reviewer: Netdata accepts request to `http`, `https` and Unix sockets, should we also specify the protocol here?

Author: Added a option to specify `protocol` as 'http' or 'https'

> - You can see how long model training is taking by looking in the logs for the collector `grep 'anomalies' /var/log/netdata/error.log` and you should see a line like `2020-10-19 12:03:34: python.d INFO: anomalies[anomalies] : training complete in 9.17 seconds (runs_counter=900, model=pca, train_n_secs=14400, models=30, n_fit_success=30, n_fit_fails=0).`. This also gives counts of the number of models, if any, that failed to fit and so had to default back to the DefaultModel (which is currently [HBOS](https://pyod.readthedocs.io/en/latest/_modules/pyod/models/hbos.html)).

Reviewer: @joelhans instead to add a warning in the middle of the paragraph, would be better to create a new line for it?

Reviewer 2: I agree!

```suggestion
- You can see how long model training is taking by looking in the logs for the collector `grep 'anomalies' /var/log/netdata/error.log` and you should see a line like `2020-10-19 12:03:34: python.d INFO: anomalies[anomalies] : training complete in 9.17 seconds (runs_counter=900, model=pca, train_n_secs=14400, models=30, n_fit_success=30, n_fit_fails=0).`. 
  -   This also gives counts of the number of models, if any, that failed to fit and so had to default back to the DefaultModel (which is currently [HBOS](https://pyod.readthedocs.io/en/latest/_modules/pyod/models/hbos.html)).
```

![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-abnormal.jpg)

Then, as the issue passes, the anomaly probabilities should settle back down into their 'normal' range again. 

![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-normal-again.jpg)

## Requirements

- This collector will only work with Python 3 and requires the packages below be installed.

```bash
# become netdata user
sudo su -s /bin/bash netdata
# install required packages for the netdata user
pip3 install --user netdata-pandas==0.0.28 numba==0.50.1 scikit-learn==0.23.2 pyod==0.8.3
```

## Configuration

Install the Python requirements above, enable the collector and restart Netdata.

```bash
cd /etc/netdata/
sudo ./edit-config python.d.conf
# Set `anomalies: no` to `anomalies: yes`

> sudo service netdata restart

Reviewer 3: What about `systemctl`? `systemd` is much more popular than `init` nowadays.

Reviewer: To complement, the complete command would be `systemctl restart netdata`

![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-normal.jpg)

If we then go onto the system and run a command like `stress-ng --all 2` to create some [stress](https://wiki.ubuntu.com/Kernel/Reference/stress-ng), we see some charts begin to have anomaly probabilities that jump outside the typical range. When the anomaly probabilities change enough, we will start seeing anomalies being flagged on the `anomalies.anomaly` chart. The idea is that these charts are the most anomalous right now so could be a good place to start your troubleshooting. 

![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-abnormal.jpg)

Then, as the issue passes, the anomaly probabilities should settle back down into their 'normal' range again. 

![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-normal-again.jpg)

## Requirements

- This collector will only work with Python 3 and requires the packages below be installed.

```bash
# become netdata user
sudo su -s /bin/bash netdata
# install required packages for the netdata user

> pip3 install --user netdata-pandas==0.0.32 numba==0.50.1 scikit-learn==0.23.2 pyod==0.8.3

Reviewer 4: `requests` package is needed

or `netdata-pandas` depends on it and it would be installed?

Author: yep - `netdata-pandas` installs it

_**Note**: If you are unsure about any of the below configuration options then it's best to just ignore all this and leave the `anomalies.conf` file alone to begin with. Then you can return to it later if you would like to tune things a bit more once the collector is running for a while and you have a feeling for its performance on your node._

Edit the `python.d/anomalies.conf` configuration file using `edit-config` from the your agent's [config
directory](https://learn.netdata.cloud/guides/step-by-step/step-04#find-your-netdataconf-file), which is usually at `/etc/netdata`.

```bash
cd /etc/netdata   # Replace this path with your Netdata config directory, if different
sudo ./edit-config python.d/anomalies.conf
```

The default configuration should look something like this. Here you can see each parameter (with sane defaults) and some information about each one and what it does.

```yaml
# Host to pull data from

> host: '127.0.0.1:19999'

Reviewer 4: Config example shoould be

```yaml
job_name: name
  param1: value1
  ...
```

Author: yeah - ill change the readme to match whats in the config so as no room for confusion.

> <!--
> ---
> title: "anomalies"
> custom_edit_url: https://github.com/netdata/netdata/edit/master/collectors/python.d.plugin/anomalies/README.md
> ---
> -->

Reviewer 2: ```suggestion
<!--
title: "Anomalies"
custom_edit_url: https://github.com/netdata/netdata/edit/master/collectors/python.d.plugin/anomalies/README.md
-->
```

> # Anomalies - ML driven anomaly detection for your Netdata!

Reviewer 2: ```suggestion
# Anomalies: ML-driven anomaly detection for the Netdata Agent
```

## Charts

Two charts are produced:

- **Anomaly Probability** (`anomalies.probability`): This chart shows the probability that the latest observed data is anomalous based on the trained model for that chart (using the [`predict_proba()`](https://pyod.readthedocs.io/en/latest/api_cc.html#pyod.models.base.BaseDetector.predict_proba) method of the trained PyOD model).
- **Anomaly** (`anomalies.anomaly`): This chart shows `1` or `0` predictions of if the latest observed data is considered anomalous or not based on the trained model (using the [`predict()`](https://pyod.readthedocs.io/en/latest/api_cc.html#pyod.models.base.BaseDetector.predict) method of the trained PyOD model).

Below is an example of the charts produced by this collector and how they might look when things are 'normal' on the node. The anomaly probabilities tend to bounce randomly around a typically low probability range, one or two might randomly jump or drift outside of this range every now and then and show up as anomalies on the anomaly chart. 

> ![alt text](https://github.com/andrewm4894/random/blob/master/images/netdata/netdata-anomalies-collector-normal.jpg)

Reviewer 2: Can you add some alt text to each of the images? It's good for SEO and accessibility purposes.

Also, I'd recommend you upload the files to GitHub's static CDN (attach them to comment and copy the URLs), so that we know they stay available and aren't directly connected to your GitHub account/repo.

Author: Will do - ill attach them in here and use that link

![netdata-alarms-collector](https://user-images.githubusercontent.com/2178292/100630484-8482c780-3322-11eb-920b-3e0018f4056e.jpg)
![netdata-anomalies-collector-abnormal](https://user-images.githubusercontent.com/2178292/100630489-851b5e00-3322-11eb-87e5-509ca63e7afd.jpg)
![netdata-anomalies-collector-normal](https://user-images.githubusercontent.com/2178292/100630492-85b3f480-3322-11eb-842a-cd3e5a5917ec.jpg)
![netdata-anomalies-collector-normal-again](https://user-images.githubusercontent.com/2178292/100630494-85b3f480-3322-11eb-94cc-9f94ced3acb1.jpg)
![netdata-anomalies-online-collector-abnormal](https://user-images.githubusercontent.com/2178292/100630497-864c8b00-3322-11eb-98c2-54a5f4d98f59.jpg)
![netdata-anomalies-online-collector-normal](https://user-images.githubusercontent.com/2178292/100630499-864c8b00-3322-11eb-9bb4-4b41c1832850.jpg)
![netdata-anomalies-online-collector-normal-again](https://user-images.githubusercontent.com/2178292/100630502-86e52180-3322-11eb-9742-de75ac8eedcd.jpg)
![netdata-zscores-collector-abnormal](https://user-images.githubusercontent.com/2178292/100630503-86e52180-3322-11eb-874c-a2f3841d4256.jpg)
![netdata-zscores-collector-normal](https://user-images.githubusercontent.com/2178292/100630504-877db800-3322-11eb-9898-cba0a3cad564.jpg)
![netdata-zscores-collector-normal-again](https://user-images.githubusercontent.com/2178292/100630506-88164e80-3322-11eb-8f81-aad63aaa0119.jpg)

```

## Configuration

Install the Python requirements above, enable the collector and restart Netdata.

```bash
cd /etc/netdata/
sudo ./edit-config python.d.conf
# Set `anomalies: no` to `anomalies: yes`
sudo systemctl restart netdata
```

The configuration for the anomalies collector defines how it will behave on your system and might take some experimentation with over time to set it optimally for your node. Out of the box, the config comes with some [sane defaults](https://www.netdata.cloud/blog/redefining-monitoring-netdata/) to get you started that try to balance the flexibility and power of the ML models with the goal of being as cheap as possible in term of cost on the node resources. 

_**Note**: If you are unsure about any of the below configuration options then it's best to just ignore all this and leave the `anomalies.conf` file alone to begin with. Then you can return to it later if you would like to tune things a bit more once the collector is running for a while and you have a feeling for its performance on your node._

Edit the `python.d/anomalies.conf` configuration file using `edit-config` from the your agent's [config

> directory](https://learn.netdata.cloud/guides/step-by-step/step-04#find-your-netdataconf-file), which is usually at `/etc/netdata`.

Reviewer 2: ```suggestion
directory](/docs/configure/nodes.md), which is usually at `/etc/netdata`.
```

    diffs_n: 1

    # What is the typical proportion of anomalies in your data on average? 
    # This paramater can control the sensitivity of your models to anomalies. 
    # Some discussion here: https://github.com/yzhao062/pyod/issues/144
    contamination: 0.001

    # Define any custom models you would like to create anomaly probabilties for, some examples below to show how.
    # For example below example creates two custom models, one to run anomaly detection on the netdata user 
    # and one on the apps metrics for python.d.plugin.
    # custom_models:
    #  - name: 'user_netdata'
    #    dimensions: 'users.cpu|netdata,users.mem|netdata,users.threads|netdata,users.processes|netdata,users.sockets|netdata'
    #  - name: 'apps_python_d_plugin'
    #    dimensions: 'apps.cpu|python.d.plugin,apps.mem|python.d.plugin,apps.threads|python.d.plugin,apps.processes|python.d.plugin,apps.sockets|python.d.plugin'

    # Set to true to normalize, using min-max standardization, features used for the custom models. 
    # Useful if your custom models contain dimensions on very different scales. 
    # Usually best to leave as false.
    # custom_models_normalize: false
```

> ## Custom Models

Reviewer 2: ```suggestion
## Custom models
```

  - Typically ~3%-3.5% cpu usage from scoring, jumping to ~50% for a couple of seconds during model training.
  - About ~150mb of ram (`apps.mem`) being continually used by the `python.d.plugin`, peaking around ~270mb for a few seconds during training using default settings.
- If you activate this collector on a fresh node, it might take a little while to build up enough data to calculate a realistic and useful model.
- Some models like `iforest` can be comparatively expensive (on same n1-standard-2 system above ~2s runtime during predict, ~40s training time, ~50% cpu on both train and predict) so if you would like to use it you might be advised to set a relativley high `update_every` maybe 10,15 or 30 in `anomalies.conf`.
- Setting a higher `train_every_n` and `update_every` is an easy way to devote less resources on the node to anomaly detection. Specifying less charts and a lower `train_n_secs` will also help reduce resources at the expense of covering less charts and maybe a more noisey model if you set `train_n_secs` to be too small for how your node tends to behave.

> ## Useful Links & Further Reading

Reviewer 2: ```suggestion
## Useful links and further reading
```
